### PR TITLE
Fix issue 16179 - multiSort no longer callable with delegate with context

### PR DESCRIPTION
**Pull request by togrue**
https://issues.dlang.org/show_bug.cgi?id=16179 (multiSort no longer callable with delegate with context)
Moved multiSort's nested functions to module scope. 

Sorting predicates that access the frame of a function should work.

### CODE REVIEW

File: std/algorithm/sorting.d [hunk 1]
```diff
@@ -846,6 +846,47 @@ private template validPredicates(E, less...)
...
+    {
+        alias lessFun = binaryFun!(f);
+        if (lessFun(a, b)) return true;
+        if (lessFun(b, a)) return false;
+    }
+    return false;
+}
+
+private void multiSortImpl(Range, SwapStrategy ss, funs...)(Range r)
+{
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: nitpick: the typical style is too put an private implementation below the public declaration, s.t. it's more obvious to a reader that might scroll top to bottom.

File: std/algorithm/sorting.d [hunk 2]
```diff
@@ -886,48 +927,8 @@ template multiSort(less...) //if (less.length > 1)
...
-                    alias lessFun = binaryFun!(f);
-                    if (lessFun(a, b)) return true;
-                    if (lessFun(b, a)) return false;
-                }
-                return false;
-            }
-            multiSortImpl!funs(r);
-            return assumeSorted!predFun(r);
+            multiSortImpl!(Range, ss, funs)(r);       
+            return assumeSorted!( multiSortPredFun!(Range, funs) )(r);
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: style nitpick: no spaces between the functions - AFAICT that's the common style in this module / Phobos

File: std/algorithm/sorting.d [hunk 3]
```diff
@@ -986,6 +987,16 @@ template multiSort(less...) //if (less.length > 1)
...
 }
 
+@safe unittest // issue 16179 (cannot access frame of function)
+{
+    int[] arr;
+    int n;
+    arr.multiSort!(
+        (a, b) => a > b,
+        (a, b) => a < n
+    );
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: if we are already at it, we could also test the execution once more:

```d
int[] arr = [[7, 20], [[7, 10], [7, 15]];
int val = 1;
arr.multiSort!(
    (a, b) => a[0] < b[0],
    (a, b) => a[1] * val < b[1] * val
);
assert(arr = [[7, 10], [[7, 15], [7, 20]]);
```

**Comment by wilzbach** [comment]
> @wilzbach: The PR was automatically assigned to you, please reassign it if you were identified mistakenly.

FYI: This feature of the dlang-bot is still in work (it tries to assign persons without merge rights).

@togrue 
1) You still seem to have an issue with trailing whitespace (If your phobos fork up-to-date, you can run `make -f posix.mak style` or `make -f posix.mak checkwhitespace`. I also recommend setting up editorconfig in your favorite editor (or switch to non-trailing whitespace by default)
2) If you change the title to "Fix 16179" or "Fix issue 16179" the dlang-bot will recognize it and automatically add it to the changelog and update the Bugzilla issue. Btw in general it's nice if you include the title of the issue in the PR too (e.g. people just reading their Github emails or browsing through the PR list)
3) I added a few nits, apart from LGTM and thanks for taking the time to fix!

File: std/algorithm/sorting.d [hunk 4]
```diff
@@ -989,13 +989,15 @@ private void multiSortImpl(Range, SwapStrategy ss, funs...)(Range r)
...
+    import std.algorithm.comparison : equal;
+
     auto arr = [[1, 2], [2, 0], [1, 0], [1, 1]];
     auto lut = [2, 3, 1];
     arr.multiSort!(
         (a, b) => a[0] < b[0],
         (a, b) => lut[a[1]] < lut[b[1]]
     );
-    assert(equals(arr, [[1, 2], [1, 0], [1, 1], [2, 0]]));
+    assert(equal(arr, [[1, 2], [1, 0], [1, 1], [2, 0]]));
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: why don't you use `==`?

Comment (togrue) [inline std/algorithm/sorting.d]: I'll change this

**Comment by wilzbach** [comment]
> 2) If you change the title to "Fix 16179" or "Fix issue 16179" the dlang-bot will recognize it and automatically add it to the changelog and update the Bugzilla issue.

Sorry it's not the PR title, but [any commit message](https://github.com/MartinNowak/dlang-bot/blob/master/source/app.d#L110) (however I [opened an enhancement request](https://github.com/MartinNowak/dlang-bot/issues/7))

**Comment by lodo1995** [comment]
@togrue can you squash your commits into one (with the "Fix issue 16179" title), so that's ready for merging?

@wilzbach if you think it's ok, this should be merged as soon as possible (after the squash), as it's blocking #4737.

**Comment by wilzbach** [comment]
> @wilzbach if you think it's ok, this should be merged as soon as possible (after the squash), as it's blocking #4737.

Oh, I lost a bit track of this one, sorry. Thanks for the ping!

> @togrue can you squash your commits into one (with the "Fix issue 16179" title), so that's ready for merging?

Yep please squash & we are good to go :)

File: std/algorithm/sorting.d [hunk 5]
```diff
@@ -986,6 +987,18 @@ template multiSort(less...) //if (less.length > 1)
...
 }
 
+@safe unittest // issue 16179 (cannot access frame of function)
+{
+    auto arr = [[1, 2], [2, 0], [1, 0], [1, 1]];
+    auto lut = [2, 3, 1];
+
+    arr.multiSort!(
+        (a, b) => a[0] < b[0],
+        (a, b) => lut[a[1]] < lut[b[1]]
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: btw this makes the test unnecessary complex, just multiplying with an integer of the function scope (as suggested in the bug report) will also trigger the frame access error

**Comment by andralex** [comment]
I'm on board with this, @wilzbach please merge when you approve.

File: std/algorithm/sorting.d [hunk 6]
```diff
@@ -886,35 +886,22 @@ template multiSort(less...) //if (less.length > 1)
             return sort!(funs[0], ss, Range)(r);
         else
         {
-            static void multiSortImpl(funs...)(Range r) @safe
-            {
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: I think you accidentally removed `multiSortImpl` with your last change?

**Comment by wilzbach** [comment]
> I'm on board with this, @wilzbach please merge when you approve.

Thanks, this is sadly solely blocked due to @togrue's problem(s) with squashing.
This problem seems to happen regularly here. Has anyone a good idea how we could ease the flow? (I already started a discussion [about auto-squashing](https://github.com/braddr/d-tester/issues/55), but this seems to be too problematic)

@togrue a couple of pointers:
https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/

**Comment by wilzbach** [comment]
@togrue I saw that you repaired your last commit. Could you please squash all your commits into one?
We would like to add only _one_ commit to the repo to keep a clean history.
Please let us know if you have troubles there - we are happy to help. A good chatting place might be #d on Freenode (you can ping me anytime - my nick is greenify) or D.learn on the NG.

**Comment by wilzbach** [comment]
Glad we could we solve the squashing problem :)

**Comment by wilzbach** [comment]
Auto-merge toggled on

**Pull request merged**